### PR TITLE
Sysupgrade backup fixes

### PR DESCRIPTION
**Pull request by luizluca**
This patch series fixes some problems with sysupgrade backup.
I divided it in three steps as independent as possible from each other.
I think it is better to understand the change this way. However, I can squash
all or any one of them if needed.

These patches should also be backported to 19.07.

### CODE REVIEW

File: package/base-files/files/sbin/sysupgrade [hunk 1]
```diff
@@ -118,25 +118,40 @@ list_conffiles() {
...
+		# in format: <space><filename><space><hash>
+		conffiles == 1 { print $(NF) "  " substr($0,2,length($0)-length($(NF))-2) }
 	' /usr/lib/opkg/status
 }
 
 list_changed_conffiles() {
-	# Cannot handle spaces in filenames - but opkg cannot either...
-	list_conffiles | while read file csum; do
+	list_conffiles | while read -r csum_file; do
+		file="$(echo "$csum_file" | cut -d" " -f3-)"
```
Comment (jow-) [inline package/base-files/files/sbin/sysupgrade]: Why is this code change needed? Why that inefficient `cut` usage? What is wrong with `read file csum` ?

Comment (luizluca) [reply package/base-files/files/sbin/sysupgrade]: > Why is this code change needed?

Current code does not support space in filename. I'm testing with an extreme case but any space at all would break sysupgrade like "/etc/my system/conf":

```
mkdir '/etc/A\ "!$HOME'"'"'  a*s?d[123] /'
touch '/etc/A\ "!$HOME'"'"'  a*s?d[123] /test_file'
echo '/etc/A\ "!$HOME'"'"'  a*s?d[123] /test_file' >>/etc/sysupgrade.conf
sysupgrade -l | grep test_file
```

I swapped file csum order to guarantee that the first matched "field separator" is not part of path. I used double space as "field separator" to match what sha256sum expects, so I do not need to build it later.

> What is wrong with `read file csum` ?

At least:
1. files will have only the first part of the filepath until it match any IFS. csum will have the remaining line.
2. any valid IFS will be converted and merged in a single space (tab, multiple spaces)
3. It is missing '-r', so \ is consumed by read

> Why that inefficient `cut` usage?

You are right. I overlooked it here. I pushed a new version that uses variable expansion "${csum_file#*  }", which is 300x faster.

File: package/base-files/files/sbin/sysupgrade [hunk 2]
```diff
@@ -60,7 +60,7 @@ export INSTALLED_PACKAGES=${ETCBACKUP_DIR}/installed_packages.txt
 
 IMAGE="$1"
 
-[ -z "$IMAGE" -a -z "$NEED_IMAGE" -a $CONF_BACKUP_LIST -eq 0 -o $HELP -gt 0 ] && {
+[ -z "$IMAGE" ] && [ -z "$NEED_IMAGE" ] && [ "$CONF_BACKUP_LIST" -eq 0 ] || [ "$HELP" -gt 0 ] && {
```
Comment (adrianschmutzler) [inline package/base-files/files/sbin/sysupgrade]: I think you should separate the cosmetic changes (like this one) from the "functional" changes (space in filename) and make separate commits for them.

Comment (luizluca) [reply package/base-files/files/sbin/sysupgrade]: The cosmetic changes were so tittle that I kept it together with the shellcheck fixes. However, should label that commit on the most important aspect (the files with space fix) and not the minor issue (shellchecks).

Anyway, I splitted the commit.

File: package/base-files/files/sbin/sysupgrade [hunk 3]
```diff
@@ -101,7 +101,7 @@ EOF
 	exit 1
 }
 
-[ -n "$IMAGE" -a -n "$NEED_IMAGE" ] && {
+[ -n "$IMAGE" ] && [ -n "$NEED_IMAGE" ] && {
```
Comment (adrianschmutzler) [inline package/base-files/files/sbin/sysupgrade:104]: This might change operator precedence.

File: package/base-files/files/sbin/sysupgrade (hunk 2 above)
Comment (adrianschmutzler) [reply package/base-files/files/sbin/sysupgrade:63]: I don't think it's a good idea to have strings in integer comparison.

**Comment by adrianschmutzler** [comment]
Since nobody seems interested, consider closing the PR and submitting to the openwrt-devel list, where you might meet more interested people.

**Comment by luizluca** [comment]
Thanks @adrianschmutzler , I even created a second PR #3045 with a subset.
